### PR TITLE
UpdateSourceTrigger

### LINKED ISSUE #3754: Porting WPF UpdateSourceTrigger
Having something like the WPF `Binding.UpdateSourceTrigger` (https://docs.microsoft.com/en-us/dotnet/api/system.windows.data.binding.updatesourcetrigger?view=netframework-4.8) in Avalonia would be helpful.

### PR DESCRIPTION
**Pull request by pr8x**
## What does the pull request do?

Implement WPF's `UpdateSourceTrigger` (https://docs.microsoft.com/en-us/dotnet/api/system.windows.data.binding.updatesourcetrigger?view=windowsdesktop-6.0)

TODO:

- Documentation
- ~~I need a proper interface for getting focus loss event inside Avalonia.Base. Any ideas? There could be some `IUpdateSourceTriggerLostFocusProvider` or something. Right now I am just checking for "IsFocused" property (by name) which is not a good solution for obvious reasons.~~

## What is the current behavior?

It's not supported.

## Checklist

- [X] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

Most important ones:
`InstancedBinding.ctor()`
`InstancedBinding.TwoWay()`
`InstancedBinding.OneWayToSource()`

These could be defaulted to `UpdateSourceTrigger.Default` though.

## Fixed issues

Fixes https://github.com/AvaloniaUI/Avalonia/issues/3754

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019276-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0019289-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by MarchingCube** [comment]
Maybe you can add an example to BindingDemo? (side note: styles are broken in there, can probably switch to fluent theme).

File: src/Avalonia.Base/Data/InstancedBinding.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Data/InstancedBinding.cs:18]: I think that you can use `SingleSubscriberObservableBase<T>` here since there should be only one subscriber from what I understand.

File: src/Avalonia.Base/Data/InstancedBinding.cs
Comment (MarchingCube) [inline src/Avalonia.Base/Data/InstancedBinding.cs:96]: Technically in WPF this works without explicit update source mode. There is also `UpdateTarget`. Both can be quite useful  to "refresh" bindings without having to raise property changed events.

**Review by MarchingCube — COMMENTED** [review]
Did a brief review. It seems `UpdateSource` might be a Pandora's box but in general having it would help writing controllable binding code.

File: src/Avalonia.Base/Data/InstancedBinding.cs
Comment (pr8x) [reply src/Avalonia.Base/Data/InstancedBinding.cs:96]: > Technically in WPF this works without explicit update source mode.

I see, I guess I can make it work outside of it.

**Comment by grokys** [comment]
> I need a proper interface for getting focus loss event inside Avalonia.Base. Any ideas? 

To be honest, I think we need to wait until https://github.com/AvaloniaUI/Avalonia/pull/5831 is merged in order to do this properly, because this feature breaks the layering we have.

File: src/Avalonia.Base/AvaloniaPropertyMetadata.cs
Comment (grokys) [inline src/Avalonia.Base/AvaloniaPropertyMetadata.cs:36]: In WPF this is called [`DefaultUpdateSourceTrigger`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkpropertymetadata.defaultupdatesourcetrigger?view=windowsdesktop-6.0) which is maybe a better name

File: src/Avalonia.Base/AvaloniaPropertyMetadata.cs
Comment (grokys) [inline src/Avalonia.Base/AvaloniaPropertyMetadata.cs:19]: Shouldn't this constructor be setting the field value? The `updateSourceTrigger` parameter seems to be unused.

File: src/Avalonia.Base/Data/BindingOperations.cs
Comment (grokys) [inline src/Avalonia.Base/Data/BindingOperations.cs:85]: As mentioned in the comments, this can be changed to be made less hacky once the core library merge is done. Maybe leave a comment here?

```suggestion
                        // HACK: Use the LostFocus event here once core libraries are merged and
						// we have access to InputElement.
                        valueObservable = Observable.FromEventPattern<AvaloniaPropertyChangedEventArgs>(
```

File: src/Avalonia.Base/Data/BindingOperations.cs
Comment (grokys) [inline src/Avalonia.Base/Data/BindingOperations.cs:64]: I'm slightly worried about the performance overhead of adding this observable to every binding, but I've been unable to profile because the many of the benchmarks now throw.

**Review by grokys — CHANGES_REQUESTED** [review]
A few comments, but unable to get much further with the review because there seems to be a problem with the PR as evidenced by the failing tests.

My main concern here will be that of performance, and adding the additional `explicitSourceUpdate` observable is probably going to add a bit of weight - especially as it uses rx operators which we shouldn't really be using at this low level due to their lousy performance.

File: src/Avalonia.Base/Data/BindingOperations.cs
Comment (pr8x) [reply src/Avalonia.Base/Data/BindingOperations.cs:85]: Are we planning to even merge this _before_ the core library merge happens?

File: src/Avalonia.Base/Data/BindingOperations.cs
Comment (pr8x) [reply src/Avalonia.Base/Data/BindingOperations.cs:64]: Yeah, we can probably write some simple (non-allocating) wrapper ourselves

File: src/Avalonia.Base/Data/BindingOperations.cs
Comment (maxkatz6) [reply src/Avalonia.Base/Data/BindingOperations.cs:85]: Core libraries are merged now

**Comment by pr8x** [comment]
Decided to remove `UpdateSource` and `UpdateTarget` for now. I feel like exposing them on the `InstancedBinding` is not really helpful in many cases and they have some overhead even when they're not used. I wonder if it's really worth having them.

___

Anyways, ran some benchmarks and it seems that this PR improves bindings a bit:

**Master**

|                          Method |       Mean |     Error |    StdDev |  Gen 0 |  Gen 1 | Allocated |
|-------------------------------- |-----------:|----------:|----------:|-------:|-------:|----------:|
|       TwoWayBinding_Via_Binding |   3.984 μs | 0.0778 μs | 0.0895 μs | 0.3395 | 0.0076 |      3 KB |
| UpdateTwoWayBinding_Via_Binding | 118.075 μs | 0.3807 μs | 0.3375 μs | 4.1504 |      - |     34 KB |

**This PR**

|                            Method |       Mean |     Error |    StdDev |  Gen 0 |  Gen 1 | Allocated |
|---------------------------------- |-----------:|----------:|----------:|-------:|-------:|----------:|
|         TwoWayBinding_Via_Binding |   3.177 μs | 0.0447 μs | 0.0396 μs | 0.3014 | 0.0076 |      2 KB |
|   UpdateTwoWayBinding_Via_Binding | 109.579 μs | 0.9755 μs | 0.9125 μs | 4.0283 |      - |     34 KB |

**Comment by grokys** [comment]
@pr8x wanted to take a look at this, but seems a bunch of binding unit tests are failing?

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0022898-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by pr8x** [comment]
@grokys UTs fixed.

**Comment by tomenscape** [comment]
@pr8x Our product is being ported from WPF and needs `UpdateSourceTrigger.Explicit`, `UpdateSource`, and `UpdateTarget`.

We can work around the lack of these options these by creating proxy properties and assigning values or bindings to them at times of our choosing, but this is an ugly hack that can cause confusion among control consumers.

Have you seen the method `AvaloniaObject.CoerceValue`? This sort of does the same thing as `UpdateTarget`, but a) only for `StyledProperty` properties and b) by recalculating the effective property value, not a specific binding.

Between the `AvaloniaObject._directBindings` and `ValueStore._values` fields, it seem to me that direct access to bindings applied to each property should be possible, allowing something equivalent to this WPF call:

```C#
myControl.GetBindingExpression(Control.MyProperty).UpdateTarget();
```

###  Use case for `UpdateSource`

We have a map pin that the user can drag around the screen. We need other UI elements to immediately respond to its position changing, but we only want to update the position value in our viewmodel (via a two-way binding on the pin) when the user lets go of the mouse.

### Use case for `UpdateTarget`

Our `TimePicker` template includes text boxes which the user can type into. If the control is in 12-hour mode and set to 11pm, then when the user types "23" into the hour field and commits the value, we need the box to change to display "11". But because the underlying `TimeSpan` value has not changed, no event is raised, the binding is not re-evaluated, and the box continues to display "23". We need to manually re-run the text box binding.

**Comment by Tenjim** [comment]
Hi guys, 
do you have news about this merge ? 

Thanks in advance.

**Comment by robloo** [comment]
Also would like to see an update on the status here. @pr8x Are you waiting for core maintainer feedback?

It would be great to get this in for 11.0.

**Comment by grokys** [comment]
I'm kinda 👎 on merging this PR right now for a couple of reasons:

1. We to do a major refactor of the binding system very soon anyway, and would like to include this feature anyway. The way it's implemented here feels rather bolted onto an architecture that wasn't really designed for it
2. There's only 1 unit test. I'm 99% sure that one unit test isn't going to cover all of the edge-cases

In the absence of more unit tests, I'd be ok with merging if it'd had extensive testing. I think I asked you a while ago @pr8x if you'd tested this in your product and you hadn't yet. Have you had chance to test it now?

**Comment by pr8x** [comment]
@grokys I did test it in one scenario and it was working fine. 

At this point this PR was stagnant waiting for review that we might as well close it and wait for the refactoring of the bindings. Current binding system seems messy anyway relying way too much on observables/LINQ selectors.

**Comment by grokys** [comment]
Ok, lets close this and I'll get started on refactoring the binding system soon.

**Comment by mightypanda** [comment]
Hello,
Any update on the possibility to have this feature (and refactoring of the binding system) in new 11 previews?
The workaround using a Behavior for lost focus seems not to work anymore.

**Pull request closed without merging**
